tools/web_features/manifest.py

```diff
@@ -0,0 +1,249 @@
+        result (WebFeaturesMap): The object to store the mapping results.
+        visited_dirs (Set[str], optional): A set of directories that have already been processed. Defaults to set().
+        prev_inherited_features (List[str], optional): A list of inherited web features from parent directories. Defaults to [].
+    """
+    for current_dir, sub_dirs, _ in os.walk(os.path.join(cmd_cfg.repo_root, rel_dir_path)):
+        # Sometimes it will add a . at the beginning. Let's resolve the absolute path to disambiguate.
+        current_dir = Path(current_dir).resolve().__str__()
+
+        # Skip if we already visited this directory
+        if current_dir in visited_dirs:
+            continue
+        visited_dirs.add(current_dir)
+
+        # Create a copy that may be built upon or cleared during this iteration.
+        inherited_features = prev_inherited_features.copy()
+
+        rel_dir_path = os.path.relpath(current_dir, cmd_cfg.repo_root)
+
+        web_feature_yml_full_path = os.path.join(current_dir, WEB_FEATURES_YML_FILENAME)
+        web_feature_file: Optional[WebFeaturesFile] = None
+        if os.path.isfile(web_feature_yml_full_path):
+            try:
+                web_feature_file = WebFeaturesFile(load_data_to_dict(
+                    open(web_feature_yml_full_path, "rb")))
+            except Exception as e:
+                raise e
+
+        WebFeatureToTestsDirMapper(
+            find_all_test_files_in_dir(cmd_cfg.repo_root, rel_dir_path, cmd_cfg.url_base),
+            web_feature_file
+        ).run(result, inherited_features)
+
+
+        for sub_dir in sub_dirs:
+            map_tests_to_web_features(
+                cmd_cfg,
+                os.path.join(rel_dir_path, sub_dir),
+                result,
+                visited_dirs,
+                inherited_features
+            )
+
+class WebFeatureManifestEncoder(json.JSONEncoder):
+    """
+    Custom JSON encoder.
+
+    WebFeaturesMap contains a dictionary where the value is of type set.
+    Sets cannot serialize to JSON by default. This encoder handles that by
+    calling WebFeaturesMap's to_dict() method.
+    """
+    def default(self, obj: Any) -> Any:
+        if isinstance(obj, WebFeaturesMap):
+            return obj.to_dict()
+        return super().default(obj)
+
+
+def main(venv: Any = None, **kwargs: Any) -> int:
+
+    assert logger is not None
+
+    repo_root = kwargs.get('repo_root') or localpaths.repo_root
+    path = kwargs.get("path") or os.path.join(repo_root, MANIFEST_FILE_NAME)
+
+    cmd_cfg = CmdConfig(repo_root, kwargs["url_base"])
+    feature_map = WebFeaturesMap()
+    map_tests_to_web_features(cmd_cfg, "", feature_map)
+    with open(path, "w") as outfile:
+        outfile.write(
+            json.dumps(
+                {
+                    "version": 1,
+                    "data": feature_map,
+                    "config": {"url_base": cmd_cfg.url_base}
```

Comment:
Is `cmd_cfg.url_base` always a relative path, or can we end up including a full path here? And do we really need "url_base", can it be used for anything once the manifest has been sent to another system like wpt.fyi?

Reply:
You're right. We can remove this. It was not used in my implementation in wpt.fyi